### PR TITLE
De-duplicate add_offset_array methods

### PR DESCRIPTION
**Pull request by jbrockmendel**
Combine DTI/TDI/PI add_offset_array and sub_offset_array methods.

Re-order `__add__` and `__sub__` if/else branches to put scalars up front.

This was going to include fixes for DTI/TDI with ndarray[timedelta64]/ndarray[datetime64], but the diff got big enough to split that off to the next one.

### CODE REVIEW

File: pandas/core/indexes/datetimelike.py [hunk 1]
```diff
@@ -637,13 +638,32 @@ def _sub_datelike(self, other):
...
+
+        warnings.warn("Adding/subtracting array of DateOffsets to "
+                      "{cls} not vectorized"
+                      .format(cls=type(self).__name__), PerformanceWarning)
+
+        res_values = op(self.astype('O').values, np.array(other))
+        kwargs = {}
+        if not is_period_dtype(self):
+            kwargs['freq'] = 'infer'
+        return self.__class__(res_values, **kwargs)
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py]: use ``.shallow_copy``

File: pandas/core/indexes/datetimelike.py [hunk 2]
```diff
@@ -637,13 +638,32 @@ def _sub_datelike(self, other):
...
+        Add or subtract array-like of DateOffset objects
 
-    def _sub_offset_array(self, other):
-        # Array/Index of DateOffset objects
-        return NotImplemented
+        Parameters
+        ----------
+        other : Index, np.ndarray
+            object-dtype containing pd.DateOffset objects
+        op : {operator.add, operator.sub}
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py]: can you assert on the operator

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/datetimelike.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: Changed it to shallow_copy and it broke several tests, so just pushed with it changed back to `self.__class__`.  For some reason `DatetimeIndex._simple_new` returns an `ndarray` if its input values have object-dtype, which is the case here.

Comment (jreback) [reply pandas/core/indexes/datetimelike.py]: so try using ``_shallow_copy_with_infer``, we by definition *never* *ever* use ``self.__class__`` in Index construction.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**